### ISSUE TITLE
Crash-Loop when powercycling or resetting the Wemos and data-wires to sdm120m are connected

### ISSUE THREAD

**Issue by Baxxy13** [nygma2004/sdm1202mqtt#1]
Hi,
first thanks for sharing your project and the nice video.

I have a new sdm120m with defaults (baudrate: 9600; id:1) and all values are zero, except voltage (of course).
The setup is actually only wired, but your pcb version 2 is on the way.

After uncomment [this line](https://github.com/nygma2004/sdm1202mqtt/blob/main/sdm1202mqtt.ino#L26) i was able to build and upload the scetch. Otherwise the error says: 
```
exit status 1
'modbusstatus' was not declared in this scope
```
I also insert my wifi and mqtt credentials in settings.h and disabled the fixed ip-setting by commenting the desired line.
My MQTT-Broker is Mosquitto as Homeassistant-AddOn.

Working:
If i flash the scetch, powercycle or reset the Wemos (via it's reset-button) **without** connected the 2 data-wires from RS-485 to TTL Converter to sdm120m, then the setup runs. Of course it says `Error: Response timed out` in the serial monitor.
If i then connect the 2 data-wires to sdm120m the communication comes up and the data will be send to the mqtt-broker.
![modbus_mqtt_explorer](https://user-images.githubusercontent.com/54898249/221879642-938c8df7-781b-4b27-b5a1-8ce4b871159a.JPG)

Not working:
If i flash the scetch, powercycle or reset the Wemos (via it's reset-button) **with** connected the 2 data-wires from RS-485 to TTL Converter to sdm120m, then the setup goes in an infintie crash-loop. I can come out of this loop when is disconnect the 2 data-wires.
If i reconnect the 2 data-wires after some seconds the setup runs then without problems.

Unfortunately i'm not familiar with the programming and debuging of such things. 
So i hope you can maybe help or give some hints.

I append 2 logfiles from the serial monitor from working and not working situation. Both starting with a reset of the Wemos via it's reset-button.
Also a screenshot from the Exception-Decoder analysis.
[working_modbus_mqtt.txt](https://github.com/nygma2004/sdm1202mqtt/files/10851081/working_modbus_mqtt.txt)
[not_working_modbus_mqtt.txt](https://github.com/nygma2004/sdm1202mqtt/files/10851082/not_working_modbus_mqtt.txt)

![modbus_mqtt_exception_decoder](https://user-images.githubusercontent.com/54898249/221878533-42787518-f2da-4fac-8825-1899be28e12c.JPG)

I have testet with 2 different Wemos D1 mini and 3 different RS-485 to TTL Converter. The Problem occurs in every constellation. 

Greetings from Berlin...
Baxxy

**Comment by nygma2004** [comment]
Hi Baxxy,

First of all, let me respond to the complication error. To be honest, I don't understand this. `globals.h` is in the same directory and unless I comment that line I get `sdm1202mqtt:26:10: fatal error: globals.h: No such file or directory` compiler error. This does not make sense, the compiler is definitely including this header file, otherwise the code would not work. But why it cannot find it is over my head. I Google-d this error but I did not find anything useful. I know that if you use double quotes in the include, that refers to a file in the project folder (where globals.h exists) and if you use `<` and `>` it refers to an installed library. Hence I don't understand the error message for me.

**Comment by Baxxy13** [comment]
Hi,
thanks for your answer.

I understand thus commenting a line with `//` that "option" is not used/included.

Thus i changed from:
`//#include "globals.h"` to `#include "globals.h"` and the scetch compiles.

I readed around the web and tested different hardware and finally i found the solution to get rid of the crash-loops.
And it was simple.

I had to additionally connect GND from the setup to port 8 [G] of the sdm120m.
(in your video you are using only port 9 [B-] and port 10 [A+] )

Since i do that i had no more crash-loops and the mqtt reconnect is also working.

I don't know if it's related to the RS-485 to TTL Converter i use. 
Or maybe it's related to a newer hardware/software revision of my sdm120m.
(it says 020204 for software and mine is labeled with SDM120M, yours SDM120)

Now i let the test-setup run and wait for the pcb's.

Thanks again.

Case closed. :wink:
